### PR TITLE
Add MPRAGE processing and visualisation

### LINKED ISSUE #43: Process MPRAGE scans
- [ ] Segmentation
- [ ] Labeling
- [ ] Coregistration of CP mode to CoV
- [ ] Extraction of values
- [ ] Creation of figure

### LINKED ISSUE #45: Add comment explaining the choice of CoV RF shimmed scans for segmentation and coregistration baseline

### PR DESCRIPTION
**Pull request by dpapp86**
Clarifies text wrt choice of CoV GRE and MPRAGE for segmentation

Fixes https://github.com/shimming-toolbox/rf-shimming-7t/issues/37
Fixes https://github.com/shimming-toolbox/rf-shimming-7t/issues/38
Fixes https://github.com/shimming-toolbox/rf-shimming-7t/issues/39
Fixes #43 
Fixes #45

### CODE REVIEW

**Comment by dpapp86** [comment]
@jcohenadad please review

This should contain everything needed to process MPRAGE scans and create the associated figure

File: data_processing.ipynb [hunk 1]
```diff
@@ -163,13 +168,14 @@
    "source": [
     "# Define useful variables\n",
     "\n",
-    "path_data = os.getcwd()\n",
+    "path_data = os.getcwd() #works if and only if this is run first and once. Otherwise it will pick up whatever the current directory is if someone wanted to rerun it\n",
```
Comment (jcohenadad) [inline data_processing.ipynb:171]: That's the motivation for #33

No need to add a comment in the notebook (better to fix #33 instead)

File: data_processing.ipynb [hunk 2]
```diff
@@ -163,13 +168,14 @@
...
     "print(f\"path_data: {path_data}\")\n",
     "path_qc = os.path.join(path_data, \"qc\")\n",
-    "shim_modes = [\"CP\", \"CoV\", \"patient\", \"phase\", \"SAReff\", \"target\", \"volume\"]\n",
+    "shim_modes = [\"CP\", \"patient\", \"volume\", \"phase\", \"CoV\", \"target\", \"SAReff\"] #reordered to match B1+ map figure\n",
+    "shim_modes_MPRAGE = [\"CP\", \"CoV\"]\n",
     "# shim_modes = [\"CP\", \"CoV\"]  # debugging\n",
     "print(f\"shim_modes: {shim_modes}\")\n",
-    "subjects = sorted(glob.glob(\"sub-*\"))\n",
+    "subjects = sorted(glob.glob(\"sub-*\")) #Same problem with rerunning it if the cwd is not the ds004906 root\n",
```
Comment (jcohenadad) [inline data_processing.ipynb:178]: https://github.com/shimming-toolbox/rf-shimming-7t/pull/46/files#r1451039834

File: data_processing.ipynb [hunk 3]
```diff
@@ -359,12 +369,12 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Extract B1+ value along the spinal cord between levels C1 and T2 (included)\n",
+    "# Extract B1+ value along the spinal cord between levels C3 and T2 (included)\n",
```
Comment (jcohenadad) [inline data_processing.ipynb:372]: This change is outside the scope of this PR.

File: data_processing.ipynb [hunk 4]
```diff
@@ -359,12 +369,12 @@
...
     "\n",
     "for subject in subjects:\n",
     "    os.chdir(os.path.join(path_data, subject, \"fmap\"))\n",
     "    for shim_mode in shim_modes:\n",
-    "        !sct_extract_metric -i {subject}_acq-{shim_mode}_TB1map.nii.gz -f {subject}_acq-anat{shim_mode}_TB1TFL_seg.nii.gz -method wa -vert 1:9 -vertfile {subject}_acq-anat{shim_mode}_TB1TFL_seg_labeled.nii.gz -perslice 1 -o TB1map_{shim_mode}.csv"
+    "        !sct_extract_metric -i {subject}_acq-{shim_mode}_TB1map.nii.gz -f {subject}_acq-anat{shim_mode}_TB1TFL_seg.nii.gz -method wa -vert 3:9 -vertfile {subject}_acq-anat{shim_mode}_TB1TFL_seg_labeled.nii.gz -perslice 1 -o TB1map_{shim_mode}.csv"
```
Comment (jcohenadad) [inline data_processing.ipynb:377]: out of scope

File: data_processing.ipynb [hunk 5]
```diff
@@ -387,7 +397,7 @@
     "x_grid = np.linspace(0, 1, 100)\n",
     "\n",
     "# z-slices corresponding to levels C1 to T2 on the PAM50 template. These will be used to scale the x-label of each subject.\n",
-    "original_vector = np.array([984, 938, 907, 870, 833, 800, 769, 735, 692, 646])\n",
+    "original_vector =np.array([907, 870, 833, 800, 769, 735, 692, 646]) # np.array([984, 938, 907, 870, 833, 800, 769, 735, 692, 646])\n",
```
Comment (jcohenadad) [inline data_processing.ipynb:400]: out of scope

File: data_processing.ipynb [hunk 6]
```diff
@@ -259,12 +269,12 @@
...
    "outputs": [],
    "source": [
-    "# Extract the signal intensity on the GRE scan within the spinal cord between levels C1 and T2 (included)\n",
+    "# Extract the signal intensity on the GRE scan within the spinal cord between levels C3 and T2 (included)\n",
     "\n",
     "for subject in subjects:\n",
     "    # TODO: loop across other shim methods    \n",
     "    os.chdir(os.path.join(path_data, subject, \"anat\"))\n",
-    "    !sct_extract_metric -i {subject}_acq-CoV_T2starw_crop.nii.gz -f {subject}_acq-CoV_T2starw_crop_seg.nii.gz -method wa -vert 1:9 -vertfile {subject}_acq-CoV_T2starw_crop_seg_labeled.nii.gz -append 1 -perslice 1 -o gre_CoV.csv\n"
+    "    !sct_extract_metric -i {subject}_acq-CoV_T2starw_crop.nii.gz -f {subject}_acq-CoV_T2starw_crop_seg.nii.gz -method wa -vert 3:9 -vertfile {subject}_acq-CoV_T2starw_crop_seg_labeled.nii.gz -append 1 -perslice 1 -o gre_CoV.csv\n"
```
Comment (jcohenadad) [inline data_processing.ipynb:277]: This change is outside the scope of this PR.

File: data_processing.ipynb [hunk 7]
```diff
@@ -259,12 +269,12 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Extract the signal intensity on the GRE scan within the spinal cord between levels C1 and T2 (included)\n",
+    "# Extract the signal intensity on the GRE scan within the spinal cord between levels C3 and T2 (included)\n",
```
Comment (jcohenadad) [inline data_processing.ipynb:272]: This change is outside the scope of this PR.

**Comment by jcohenadad** [comment]
> also adresses #38

that should go in the PR body-- i've added it

File: data_processing.ipynb (hunk 6 above)
Comment (jcohenadad) [reply data_processing.ipynb:276]: Not anymore

**Review by jcohenadad — COMMENTED** [review]
on https://github.com/shimming-toolbox/rf-shimming-7t/pull/46/commits/36bb2d7a4f6d8d3317eae706d26f713a79386b71 I'm getting:

```console
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[9], line 8
      6 os.chdir(os.path.join(path_data, subject, "anat"))
      7 CoV_flatten=glob.glob(os.path.join(path_data, subject, "anat", f"{subject}_acq-CoV_T1w_crop_flatten.nii.gz"))
----> 8 CoV_flatten=CoV_flatten[0]
      9 CoV_flatten_data=nib.load(CoV_flatten).get_fdata()
     10 CP_flatten=glob.glob(os.path.join(path_data, subject, "anat", f"{subject}_acq-CP_T1w_reg_flatten.nii.gz"))

IndexError: list index out of range
```

False alarm: I forgot to run required cells before :🤦

**Review by jcohenadad — CHANGES_REQUESTED** [review]
Vertebral labeling failing for sub-02 and sub-04:

Screenshot 2024-01-15 at 10 15 26 PM

Screenshot 2024-01-15 at 10 15 28 PM

**Review by jcohenadad — APPROVED** [review]

**Pull request merged**